server/controllers/project.controller.js

```diff
@@ -175,22 +203,43 @@ export function getProjects(req, res) {
 
 export function getProjectsForUser(req, res) {
   if (req.params.username) {
-    User.findOne({ username: req.params.username }, (err, user) => {
-      if (!user) {
-        res.status(404).json({ message: 'User with that username does not exist.' });
-        return;
-      }
-      Project.find({ user: user._id })
-        .sort('-createdAt')
-        .select('name files id createdAt updatedAt')
-        .exec((innerErr, projects) => res.json(projects));
-    });
+    getProjectsForUserName(req.params.username)
+      .then(projects => res.json(projects))
+      .catch((err) => {
+        if (err instanceof UserNotFoundError) {
+          res.status(404).json({ message: 'User with that username does not exist.' });
+        } else {
+          res.status(500).json({ message: 'Error fetching projects' });
+        }
+      });
   } else {
     // could just move this to client side
     res.json([]);
   }
 }
 
+export function apiGetProjectsForUser(req, res) {
```

Comment:
I don't really like duplicating these functions as they're so similar, but they return different objects.

Reply:
agreed, though it feels like the way to merge them is awkward/possibly confusing.

Reply:
Yeah, I think that's true. Where possible, I've been extracting the shared data fetching logic into a helper function that can be called from both. And then the API controller functions can be used to format the data into a specific JSON structure.